File: backend/app/schemas/demand.py
# backend/app/schemas/demand.py
from datetime import date

from pydantic import BaseModel, field_validator

DEMAND_STATUSES = ["pendente", "em_andamento", "em_aprovacao", "reprovada", "concluida"]


def _validate_status(value: str) -> str:
    if value not in DEMAND_STATUSES:
        raise ValueError(f"Status inválido. Use um de: {', '.join(DEMAND_STATUSES)}")
    return value


def _non_empty(value: str, field_label: str) -> str:
    value = value.strip()
    if not value:
        raise ValueError(f"{field_label} é obrigatório")
    return value


class ChecklistItem(BaseModel):
    id: str
    text: str
    done: bool = False


class DemandCreate(BaseModel):
    client_id: int | None = None
    title: str
    notes: str | None = None
    assignee_id: int | None = None
    is_art: bool = False
    status: str = "pendente"
    due_date: date | None = None
    has_material: bool = False
    checklist: list[ChecklistItem] = []

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: str) -> str:
        return _non_empty(value, "Título")

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        return _validate_status(value)


class DemandUpdate(BaseModel):
    client_id: int | None = None
    title: str | None = None
    notes: str | None = None
    assignee_id: int | None = None
    status: str | None = None
    due_date: date | None = None
    has_material: bool | None = None
    checklist: list[ChecklistItem] | None = None

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: str | None) -> str | None:
        return _non_empty(value, "Título") if value is not None else None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str | None) -> str | None:
        return _validate_status(value) if value is not None else None


class DemandOut(BaseModel):
    id: int
    client_id: int | None
    calendar_entry_id: int | None
    title: str
    notes: str | None
    requester_id: int
    assignee_id: int | None
    is_art: bool
    status: str
    due_date: date | None
    has_material: bool
    checklist: list[ChecklistItem]

    model_config = {"from_attributes": True}

File: crm/types/demand.ts
export const DEMAND_STATUSES = ["pendente", "em_andamento", "em_aprovacao", "reprovada", "concluida"] as const;
export type DemandStatus = (typeof DEMAND_STATUSES)[number];

export const DEMAND_STATUS_LABELS: Record<DemandStatus, string> = {
  pendente: "Pendente",
  em_andamento: "Em andamento",
  em_aprovacao: "Em aprovação",
  reprovada: "Reprovada",
  concluida: "Concluída",
};

export const DEMAND_STATUS_COLORS: Record<DemandStatus, string> = {
  pendente: "bg-white/10 text-white/70",
  em_andamento: "bg-amber-400/15 text-amber-400",
  em_aprovacao: "bg-purple-400/15 text-purple-400",
  reprovada: "bg-red-400/15 text-red-400",
  concluida: "bg-emerald-400/15 text-emerald-400",
};

export type ChecklistItem = {
  id: string;
  text: string;
  done: boolean;
};

export type Demand = {
  id: number;
  client_id: number | null;
  calendar_entry_id: number | null;
  title: string;
  notes: string | null;
  requester_id: number;
  assignee_id: number | null;
  is_art: boolean;
  status: DemandStatus;
  due_date: string | null;
  /** "Já tem material?" — já existe imagem/vídeo pronto, ou precisa captar do zero. */
  has_material: boolean;
  checklist: ChecklistItem[];
};

export type DemandCreateInput = {
  client_id?: number | null;
  title: string;
  notes?: string;
  assignee_id?: number | null;
  is_art?: boolean;
  status?: DemandStatus;
  due_date?: string | null;
  has_material?: boolean;
  checklist?: ChecklistItem[];
};

export type DemandUpdateInput = Partial<{
  client_id: number | null;
  title: string;
  notes: string;
  assignee_id: number | null;
  status: DemandStatus;
  due_date: string | null;
  has_material: boolean;
  checklist: ChecklistItem[];
}>;

export type DemandFilters = Partial<{
  client_id: number;
  assignee_id: number;
  /** Mutuamente exclusivo com `assignee_id` — "sem responsável". */
  unassigned: boolean;
  status: DemandStatus;
  date_from: string;
  date_to: string;
}>;

/** Atrasada = tem prazo, já passou, e ainda não foi concluída. */
export function isDemandOverdue(demand: Demand): boolean {
  if (!demand.due_date || demand.status === "concluida") return false;
  return demand.due_date < new Date().toISOString().slice(0, 10);
}

export type SituacaoGeral = "Atrasada" | "No prazo" | "Concluída" | "Sem prazo";

export const SITUACAO_COLORS: Record<SituacaoGeral, string> = {
  Atrasada: "bg-red-400/15 text-red-400 border border-red-400/30",
  "No prazo": "bg-cyan/15 text-cyan border border-cyan/30",
  Concluída: "bg-emerald-400/15 text-emerald-400 border border-emerald-400/30",
  "Sem prazo": "bg-white/10 text-white/50 border border-white/10",
};

function pad(n: number) {
  return String(n).padStart(2, "0");
}

/** Diferença legível entre agora e o prazo (ex: "21h 46min"). */
function formatDuration(ms: number) {
  const totalMin = Math.round(Math.abs(ms) / 60_000);
  const h = Math.floor(totalMin / 60);
  const m = totalMin % 60;
  if (h === 0) return `${m}min`;
  if (m === 0) return `${h}h`;
  return `${h}h ${pad(m)}min`;
}

/** Situação em tempo real da demanda — sem completed_at, "concluída" não distingue se foi no prazo. */
export function getDemandSituacao(demand: Demand, now = new Date()): { specific: string; general: SituacaoGeral } {
  if (demand.status === "concluida") {
    return { specific: "Concluída", general: "Concluída" };
  }
  if (!demand.due_date) {
    return { specific: "Sem prazo definido", general: "Sem prazo" };
  }

  const due = new Date(`${demand.due_date}T23:59:59`);
  const diff = due.getTime() - now.getTime();

  return diff >= 0
    ? { specific: `Dentro do prazo por ${formatDuration(diff)}`, general: "No prazo" }
    : { specific: `Atrasada há ${formatDuration(diff)}`, general: "Atrasada" };
}

export function checklistProgress(demand: Demand): { done: number; total: number } {
  return { done: demand.checklist.filter((item) => item.done).length, total: demand.checklist.length };
}
